### ISSUE TITLE
Token Space Efficiency Enhancement

### ISSUE THREAD

**Issue by lollerfirst** [cashubtc/nuts#140]
I'd like to discuss an improvement that would enhance space efficiency in tokens.

From NUT-00, a token is composed like this:
```
{
  "token": [
    {
      "mint": str,
      "proofs": Array<Proof>
    },
    ...
  ],
  "unit": str <optional>,
  "memo": str <optional>
}
```
Where the bulk of it is in `"proofs"` which is a ```Array<Proof>``` and `Proof`:
```
{
  "amount": int, 
  "id": hex_str,
  "secret": str,
  "C": hex_str
}
```
I note that it is redundant to include a "C" for every proof. Instead -when generating a token- the wallet could sum up all the C of all the proofs the user intends to send:
```C_tot = C_0 + C_1 + ... + C_n```
Then we could have that the "proofs" field in the token structure is not anymore a `Array<Proof>` but something like `AggregateProof`:
```
{
  "parts": Array<Proof>,
  "C_tot": C_tot
}
```
And `Proof` modified as follows:
```
{
  "amount": int, 
  "id": hex_str,
  "secret": str
}
```
For the receiver of this new token, nothing changes: he will still send these proofs to the mint to have them molten or swapped.

The mint, however, would have to slightly adjust its way of verifying the proofs:
```
k_0 * hash_to_curve(x_0) + k_1 * hash_to_curve(x_1) + ... + k_n * hash_to_curve(x_n) == C_tot
```
where `k_*` are the keys for the respective amounts and denominations, and `x_*` are the secrets.

Thoughts?

**Comment by lollerfirst** [comment]
The main disadvantage of this hack is the inability to compose the DLEQ proofs the same way. That means it cannot be used offline.

**Comment by conduition** [comment]
I benchmarked a few different approaches to proof aggregation in Rust using `libsecp256k1` bindings, and curve point addition actually turned out to be faster than hashing, regardless of how many proofs are being verified. Any claims we make about performance should always be done on a per-implementation basis, but I think in general point addition is a very cheap operation across most implementations so using that for the spec seems reasonable to me.

Point addition also has the advantage of being aggregatable: A client could receive, say, 10 tokens separately, and aggregate all ten `C_tot` values together in a single swap request to claim them. If the swap request fails, they can retry each individually to root out the invalid one. Hashes cannot be aggregated and so the receiver would always need 10 separate swap requests to claim all 10 tokens (or else we'd need to modify NUT-03 in an undesirable way to support it)


@lollerfirst is correct regarding offline verification. This seems like it should be a relatively simple thing for clients though: Proof aggregation or DLEQ, your tokens can use one but not the other.

**Comment by lollerfirst** [comment]
### Update on this
Just found out that:
* Solution number 1 ($C_t = C_0 + C_1 + ... + C_n$) leads to forgery. A client can request a blind signature on the following message $B' = H(x) + H(x') + r\cdot G$. After unblinding they are left with $C = a\cdot (H(x) + H(x'))$. Now they can just present $C_t = C$ and $x, x'$ individually.
* Solution number 2 ( $C_t = H(C_0 \mathbin{\Vert} C_1 \mathbin{\Vert} ... \mathbin{\Vert} C_n $) is still good, assuming the hashing function is sufficiently collision-resistant.

**Comment by conduition** [comment]
Great find. There's probably a way to make point addition safe by randomizing the points, but why risk it? Hashing would be the safer solution, as well as being a little faster.

If we care about aggregating tokens together, we could define `C_tot` as a merkle tree root hash, and structure the proofs into a merkle tree. For example, given 5 proofs `[C_1 .. C_5]`, we'd compute `C_tot` as:

```py
C_tot = sha256(
  sha256(C_1 || C_2),
  sha256(
    sha256(C_3 || C_4),
    C_5
  )
)
```

In tokens and swap requests, the client gives the structure of the tree using arrays/tuples:

```json
[
  [
    {
      "amount": 2,
      "secret": "12abcd..."
    },
    {
      "amount": 4,
      "secret": "12abcd..."
    }
  ],
  [
    [
      {
        "amount": 8,
        "secret": "12abcd..."
      },
      {
        "amount": 16,
        "secret": "12abcd..."
      }
    ],
    {
      "amount": 32,
      "secret": "12abcd..."
    }
  ],
]
```

to avoid unneeded repetition, the individual objects in the tree should just be tuples `[amount, secret]`. The above example is verbose only for clarity. Keyset IDs should likewise be deduplicated and housed at a higher level in the token.

Such structured proofs could be easily aggregated together while offline, and spent in a single request by merging the merkle trees together.

**Comment by lollerfirst** [comment]
> There's probably a way to make point addition safe by randomizing the points

Yes there is and there is another ecash protocol (Fedimint) that cannot use this solution for aggregate verification, because they use pairing crypto. In their case, they're not interested in any space efficiency, but in the fact that they could use only 1 pairing operation to verify a bunch of signatures. I wrote a gist about it if you're interested.
